### PR TITLE
Fix x ranges upon projection

### PR DESCRIPTION
**Pull request by ahuang11**
Data no longer disappears upon zoom as described in https://github.com/holoviz/geoviews/issues/673#issuecomment-2050674997.

Closes https://github.com/holoviz/geoviews/issues/673

https://github.com/holoviz/geoviews/assets/15331990/9da07622-f4a6-4d63-bdce-abc27b8962b9

```python
import xarray as xr
import panel as pn
import holoviews as hv
from holoviews.operation.datashader import rasterize
import geoviews as gv
import cartopy.crs as ccrs

hv.renderer("bokeh").webgl = False
import hvplot.xarray
ds = xr.tutorial.open_dataset("air_temperature").isel(time=0)

rasterize((gv.Image(ds, ["lon", "lat"], ["air"]))) * gv.feature.coastline()
```

### CODE REVIEW

File: geoviews/plotting/bokeh/callbacks.py [hunk 1]
```diff
@@ -118,7 +122,8 @@ class GeoRangeXYCallback(RangeXYCallback):
 
     def _process_msg(self, msg):
         msg = super()._process_msg(msg)
-        return project_ranges(self, msg, ('x_range', 'y_range'))
+        ranges = project_ranges(self, msg, ('x_range', 'y_range'))
+        return ranges
```
Comment (jbednar) [inline geoviews/plotting/bokeh/callbacks.py:126]: Doesn't seem to add anything?

**Review by jbednar — COMMENTED** [review]
Thanks for addressing this. I can't quite decide if it's the right fix, though; would be good to have e.g. @philippjfr think about it.

File: geoviews/plotting/bokeh/callbacks.py (hunk 1 above)
Comment (ahuang11) [reply geoviews/plotting/bokeh/callbacks.py:126]: Oh yeah, I can revert that; it was mainly to printing it out during debugging.

Comment (ahuang11) [inline geoviews/plotting/bokeh/callbacks.py:126]:
```suggestion
        return project_ranges(self, msg, ('x_range', 'y_range'))
```

**Comment by ahuang11** [comment]
My other two thoughts about approaches to fixing this is:

1. require -180 to 180 for PlateCarree, else raise an error
2. same as 1, but raise a warning instead
3. if 0-360 is detected, automatically convert the data (potentially costly to load data)

This PR's fix only updates two integers, which is basically "free"

File: geoviews/plotting/bokeh/plot.py [hunk 2]
```diff
@@ -70,6 +70,11 @@ def __init__(self, element, **params):
                     'and disable the show_grid option.'
                 )
 
+        self._unwrap_lons = False
+        if isinstance(self.geographic, _CylindricalProjection):
+            x1, x2 = element.range(0)
```
Comment (philippjfr) [inline geoviews/plotting/bokeh/plot.py:75]: I think element here is likely misnamed, we probably have to move this into some other method, e.g. `initialize_plot` and `update_frame`.

**Comment by TheoMathurin** [comment]
Maybe this can possibly also fix #593?

**Comment by ahuang11** [comment]
Maybe! If it uses lat/lon, then there's a chance.

File: geoviews/plotting/bokeh/plot.py [hunk 3]
```diff
@@ -113,8 +120,15 @@ def initialize_plot(self, ranges=None, plot=None, plots=None, source=None):
                                      overlaid=True, renderer=self.renderer)
             shapeplot.geographic = False
             shapeplot.initialize_plot(plot=fig)
+        nonempty = [(k, el) for k, el in self.hmap.data.items() if el]
+        self._set_unwrap_lons(nonempty[-1][1])
         return fig
 
+    def update_frame(self, key, ranges=None, element=None):
+        if element:
```
Comment (philippjfr) [inline geoviews/plotting/bokeh/plot.py:128]:
```suggestion
        if element is not None:
```

File: geoviews/plotting/bokeh/plot.py [hunk 4]
```diff
@@ -113,8 +120,15 @@ def initialize_plot(self, ranges=None, plot=None, plots=None, source=None):
                                      overlaid=True, renderer=self.renderer)
             shapeplot.geographic = False
             shapeplot.initialize_plot(plot=fig)
+        nonempty = [(k, el) for k, el in self.hmap.data.items() if el]
+        self._set_unwrap_lons(nonempty[-1][1])
```
Comment (philippjfr) [inline geoviews/plotting/bokeh/plot.py:124]:
```suggestion
        self._set_unwrap_lons(self.current_frame)
```

**Comment by ahuang11** [comment]
The changes work when it's not overlaid with coastline.

```python
import xarray as xr
import panel as pn
import holoviews as hv
from holoviews.operation.datashader import rasterize
import geoviews as gv
import cartopy.crs as ccrs

hv.extension("bokeh")
import hvplot.xarray
ds = xr.tutorial.open_dataset("air_temperature").isel(time=0)

rasterize(gv.Image(ds, ["lon", "lat"], ["air"]))
```

However, once coastline is overlaid, it produces a blank image.
image

File: geoviews/plotting/bokeh/callbacks.py [hunk 5]
```diff
@@ -64,6 +64,10 @@ def project_ranges(cb, msg, attributes):
     extents = x0, y0, x1, y1
     x0, y0, x1, y1 = project_extents(extents, plot.projection,
                                      plot.current_frame.crs)
+    if plot._unwrap_lons and -180 <= x0 < 0 or -180 <= x1 < 0:
```
Comment (philippjfr) [inline geoviews/plotting/bokeh/callbacks.py:67]: This should be:

```
if plot._unwrap_lons and (-180 <= x0 < 0 or -180 <= x1 < 0):
```

**Pull request merged**